python/cudf/cudf/core/series.py

```diff
@@ -960,23 +956,61 @@ def memory_usage(self, index=True, deep=False):
         return sum(super().memory_usage(index, deep).values())
 
     def __array_function__(self, func, types, args, kwargs):
-        handled_types = [cudf.Series]
-        for t in types:
-            if t not in handled_types:
+        if "out" in kwargs or any(not issubclass(t, (Series,)) for t in types):
+            return NotImplemented
+
+        fname = func.__name__
+
+        try:
+            # Apply a Series method if one exists.
+            cudf_ser_func = getattr(Series, fname, None)
+            if cudf_ser_func:
+                return cudf_ser_func(*args, **kwargs)
+
+            # Assume that cupy subpackages match numpy and search the
+            # corresponding cupy submodule based on the func's __module__.
+            numpy_submodule = func.__module__.split(".")[1:]
+            cupy_submodule = getattr(cupy, ".".join(numpy_submodule), None)
+            cupy_func = cupy_submodule and getattr(cupy_submodule, fname, None)
```

Comment:
How does `getattr(cupy, ".".join(numpy_submodule))` work? `getattr` can't fetch attributes with dots in them as far as I can tell from local experiments. If that's correct, this may only work if `len(numpy_submodule) == 1`. Do you need to get an attribute recursively? Moreover, can you merge the definitions of `cupy_submodule` and `cupy_func`? The value of `cupy_submodule` is only used as a pre-condition for defining `cupy_func` as something that isn't falsy. It seems as though you need something like the following:
```suggestion
            cupy_func = cupy
            for name in (*numpy_submodule, func.__name__):
                cupy_func = getattr(cupy_func, name, None)
```

Reply:
Local experiment:

```python
>>> class A:
>>>     pass
>>> a = A()
>>> a.b = A()
>>> a.b.c = "Hello"
>>> getattr(a, "b.c")
Traceback [...]
AttributeError: 'A' object has no attribute 'b.c'
```

Reply:
I guess this was missed in test coverage? Would be good to add tests covering it if possible.

Reply:
Can you think of any such functions? I can't think of any functions off the top of my head that are that deeply nested in numpy. You would need something like `numpy.subpackage1.subpackage2.func`, and you would need that function to support the array function protocol.

Reply:
Not many functions are deeply nested, as you point out. Is this a possible example? https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.polyfit.html

The part I’m not sure about is which functions are (or could be) part of the array function protocol. I could imagine calling polyfit with cupy arrays as arguments and dispatching  accordingly to a GPU implementation. I haven’t read the source to see if this actually has such a dispatch.

Reply:
This one is not:
```
>>> import numpy as np
>>> class A:
...     def __array_function__(self, func, types, args, kwargs):
...             print("In array function")
... 
>>> np.linalg.norm(A())
In array function
>>> np.polynomial.polynomial.polyfit(A(), A(), 3)
Traceback (most recent call last):
...
TypeError: unsupported operand type(s) for +: 'A' and 'float'
```

Unfortunately AFAICT there is no way to know which functions are part of the array function protocol. My understanding is that any function that is not a ufunc _could be_, but there's no way to know for sure whether one _is_ without trying.

I think we should just let this slide until someone finds an issue with it. The reality is that this implementation of `__array_function__` will almost certainly fail for a large chunk of the numpy API that supports the protocol. Most numpy methods beyond the simplest (e.g. reductions) would require specialized implementations (think `np.stack` or `np.concatenate`), but users of cuDF have indicated that it's more useful to have this dispatch work whenever it works rather than employing an allowlist (which is what [NEP 18 actually recommends](https://numpy.org/neps/nep-0018-array-function-protocol.html#example-for-a-project-implementing-the-numpy-api)). As a result, I think we just accept that our coverage right now is based on numpy functions that people have cared about in the past, and if we encounter others that fail we can add them to our test suite.

Reply:
Good enough for me.